### PR TITLE
[feat] Add Lovasz-Hinge/Softmax losses

### PR DESCRIPTION
**Pull request by ChristophReich1996**
#### Changes

Adding two new losses for semantic segmentation ([Lovasz-Hinge](http://proceedings.mlr.press/v37/yub15.pdf) (binary) and [Lovasz-Softmax](https://arxiv.org/pdf/1705.08790.pdf) (multi-class) loss). Implementation is based on the authors implementation of the Lovasz-Softmax loss ([see](https://github.com/bermanmaxim/LovaszSoftmax)).

Both loss functions are widely used for semantic segmentation tasks and provide a differentiable surrogate of the intersection-over-union (IoU) enabling direct optimization of the IoU.

#### Type of change

- [x] 📚  Documentation Update
- [x] 🧪 Tests Cases
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🔬 New feature (non-breaking change which adds functionality)
- [ ] 🚨 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📝 This change requires a documentation update

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

**Comment by edgarriba** [comment]
@ChristophReich1996 thanks so much for the contribution. I’ll submit a review tomorrow ASAP.

File: kornia/losses/lovasz_hinge.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        >>> output = lovasz_hinge_loss(input, target)
+        >>> output.backward()
+    """
+    if not isinstance(input, torch.Tensor):
+        raise TypeError(f"Input type is not a torch.Tensor. Got {type(input)}")
+
+    if not isinstance(target, torch.Tensor):
+        raise TypeError(f"Target type is not a torch.Tensor. Got {type(target)}")
+
+    if not len(input.shape) == 4:
```
Comment (edgarriba) [inline kornia/losses/lovasz_hinge.py:58]: use `KORNIA_CHECK_SHAPE(input, ["B", "1", "H", "W"])`

same for the others, and explore the testing utilities
https://github.com/kornia/kornia/blob/master/kornia/testing/__init__.py#L186

File: kornia/losses/lovasz_hinge.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+    # get shapes
+    B, N = input_flatten.shape
+
+    # compute probabilities
+    input_prob: torch.Tensor = torch.sigmoid(input_flatten)
+
+    # compute actual loss
+    signs = 2. * target_flatten - 1.
+    errors = 1. - input_prob * signs
```
Comment (edgarriba) [inline kornia/losses/lovasz_hinge.py:85]:
```suggestion
    errors = 1. - logits * signs
```
as the original implementation ?

File: kornia/losses/lovasz_hinge.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+    if not input.shape[-2:] == target.shape[-2:]:
+        raise ValueError(f"input and target shapes must be the same. Got: {input.shape} and {target.shape}")
+
+    if not input.device == target.device:
+        raise ValueError(f"input and target must be in the same device. Got: {input.device} and {target.device}")
+
+    # flatten input and target [B, -1] and to float
+    input_flatten: torch.Tensor = input.flatten(start_dim=1)
+    target_flatten: torch.Tensor = target.flatten(start_dim=1).float()
```
Comment (edgarriba) [inline kornia/losses/lovasz_hinge.py:75]:
```suggestion
    target_flatten: torch.Tensor = target.flatten(start_dim=1)
```
we should make generic to any data type

File: kornia/losses/lovasz_hinge.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        raise ValueError(f"Invalid input shape, we expect Bx1xHxW. Got: {input.shape}")
+
+    if not input.shape[-2:] == target.shape[-2:]:
+        raise ValueError(f"input and target shapes must be the same. Got: {input.shape} and {target.shape}")
+
+    if not input.device == target.device:
+        raise ValueError(f"input and target must be in the same device. Got: {input.device} and {target.device}")
+
+    # flatten input and target [B, -1] and to float
+    input_flatten: torch.Tensor = input.flatten(start_dim=1)
```
Comment (edgarriba) [inline kornia/losses/lovasz_hinge.py:74]:
```suggestion
    input_flatten: Tensor = input.reshape(B, -1)
```
1. We are trying to isolate torch imports to make the api more pythonic and generic potentially to other frameworks. I'd recommend to `from torch import Tensor` and possibly reduce the Tensor typings to make more readable the code. Specially for losses which are mostly tensors and includes mathematical formulation.
2. Use `.reshape()` since it's used across different frameworks, like numpy, jax, etc

File: kornia/losses/lovasz_hinge.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    target_flatten: torch.Tensor = target.flatten(start_dim=1).float()
+
+    # get shapes
+    B, N = input_flatten.shape
+
+    # compute probabilities
+    input_prob: torch.Tensor = torch.sigmoid(input_flatten)
```
Comment (edgarriba) [inline kornia/losses/lovasz_hinge.py:81]: can we operate directly on the logits ? in the original implementation code they seems make that assumption

File: kornia/losses/lovasz_hinge.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+    # compute actual loss
+    signs = 2. * target_flatten - 1.
+    errors = 1. - input_prob * signs
+    errors_sorted, permutation = torch.sort(errors, dim=1, descending=True)
```
Comment (edgarriba) [inline kornia/losses/lovasz_hinge.py:86]:
```suggestion
    errors_sorted, permutation = errors.sort(dim=1, descending=True)
```
to avoid torch imports

File: kornia/losses/lovasz_hinge.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    batch_index: torch.Tensor = torch.arange(B, device=input.device).repeat_interleave(N, dim=0)
```
Comment (edgarriba) [inline kornia/losses/lovasz_hinge.py:87]: `repeat_interleave` can be achieved with reshape and repeat

File: kornia/losses/lovasz_hinge.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+    target_sorted: torch.Tensor = target_flatten[batch_index, permutation.view(-1)]
+    target_sorted: torch.Tensor = target_sorted.view(B, N)
+    target_sorted_sum: torch.Tensor = target_sorted.sum(dim=1, keepdim=True)
+    intersection: torch.Tensor = target_sorted_sum - target_sorted.cumsum(dim=1)
```
Comment (edgarriba) [inline kornia/losses/lovasz_hinge.py:91]:
```suggestion
    intersection: torch.Tensor = target_sorted_sum - target_sorted.cumsum(1)
```
pytorch is `dim` others is `axis`

File: kornia/losses/lovasz_hinge.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+    union: torch.Tensor = target_sorted_sum + (1. - target_sorted).cumsum(dim=1)
+    gradient: torch.Tensor = 1. - intersection / union
+    if N > 1:
+        gradient[..., 1:] = gradient[..., 1:] - gradient[..., :-1]
+    loss: torch.Tensor = (F.relu(errors_sorted) * gradient).sum(dim=1).mean()
```
Comment (edgarriba) [inline kornia/losses/lovasz_hinge.py:96]:
```suggestion
    loss: torch.Tensor = (errors_sorted.relu() * gradient).sum(dim=1).mean()
```
or even `clip`, `clamp` but again pytorch vs the rest

File: kornia/losses/lovasz_softmax.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+
+    # flatten input [B, C, -1] and target [B, -1] and to float
+    input_flatten: torch.Tensor = input.flatten(start_dim=2)
+    target_flatten: torch.Tensor = target.flatten(start_dim=1).float()
+
+    # get shapes
+    B, C, N = input_flatten.shape
+
+    # compute softmax over the classes axis
+    input_soft: torch.Tensor = F.softmax(input_flatten, dim=1)
```
Comment (edgarriba) [inline kornia/losses/lovasz_softmax.py:83]:
```suggestion
    input_soft: torch.Tensor = input_flatten.softmax(1)
```

File: kornia/losses/lovasz_softmax.py [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+
+    # compute actual loss
+    losses: List[torch.Tensor] = []
+    batch_index: torch.Tensor = torch.arange(B, device=input.device).repeat_interleave(N, dim=0)
+    for c in range(C):
+        foreground: torch.Tensor = (target_flatten == c).float()
```
Comment (edgarriba) [inline kornia/losses/lovasz_softmax.py:89]:
```suggestion
        foreground: torch.Tensor = (target_flatten == c)
```

**Review by edgarriba — COMMENTED** [review]
Mostly comments about how to make code more readable and general to other than pytorch. In general LGMT -- check also the failures regarding the typing redefinition and docstrings

File: kornia/losses/lovasz_hinge.py (hunk 5 above)
Comment (ChristophReich1996) [reply kornia/losses/lovasz_hinge.py:81]: Yes, we can do this, I thought applying sigmoid/softmax inside the loss function module is more consistent with the [Dice loss](https://github.com/kornia/kornia/blob/42c5a88453d5b79054c2a0b5c826b4586ba505bb/kornia/losses/dice.py#L62) and the [Focal loss](https://github.com/kornia/kornia/blob/42c5a88453d5b79054c2a0b5c826b4586ba505bb/kornia/losses/focal.py#L81) implementation.

File: kornia/losses/lovasz_hinge.py (hunk 4 above)
Comment (ChristophReich1996) [reply kornia/losses/lovasz_hinge.py]: What's about the .view() call, should I go also for .reshape() there? If not done!

File: kornia/losses/lovasz_hinge.py (hunk 7 above)
Comment (ChristophReich1996) [reply kornia/losses/lovasz_hinge.py]: Done!

File: kornia/losses/lovasz_hinge.py (hunk 2 above)
Comment (ChristophReich1996) [reply kornia/losses/lovasz_hinge.py]: The original code is a bit misleading, the input to the Lovasz-Hinge loss is a tensor with values in the range (-inf, inf), not logits! I fixed this in my code.

File: kornia/losses/lovasz_hinge.py (hunk 1 above)
Comment (ChristophReich1996) [reply kornia/losses/lovasz_hinge.py]: For checking the shapes `KORNIA_CHECK_SHAPE` is used. However, the other checks are still performed "by hand" because `KORNIA_CHECK` does not work with `torch.jit`...

**Review by edgarriba — APPROVED** [review]
LGTM
